fxrunner/src/lib/osapi.rs

```diff
@@ -0,0 +1,157 @@
+    ///
+    /// The returned value is between 0 and 1.
+    fn get_cpu_idle_time(&self) -> Result<f64, Self::CpuTimeError>;
+}
+
+/// A [`ShutdownProvider`](trait.ShutdownProvider.html) that uses the Windows API.
+#[derive(Debug, Default)]
+pub struct WindowsShutdownProvider {
+    /// Whether or not to skip the actual restart.
+    #[cfg(debug_assertions)]
+    skip_restart: bool,
+}
+
+#[cfg(debug_assertions)]
+impl WindowsShutdownProvider {
+    pub fn skipping_restart(skip_restart: bool) -> Self {
+        let mut provider = WindowsShutdownProvider::default();
+        provider.skip_restart = skip_restart;
+        provider
+    }
+}
+
+impl ShutdownProvider for WindowsShutdownProvider {
+    type Error = shutdown::ShutdownError;
+
+    #[cfg(debug_assertions)]
+    fn initiate_restart(&self, reason: &str) -> Result<(), Self::Error> {
+        if self.skip_restart {
+            Ok(())
+        } else {
+            shutdown::initiate_restart(reason)
+        }
+    }
+
+    #[cfg(not(debug_assertions))]
+    fn initiate_restart(&self, reason: &str) -> Result<(), Self::Error> {
+        shutdown::initiate_restart(reason)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct WindowsPerfProvider;
+
+impl PerfProvider for WindowsPerfProvider {
+    type DiskIoError = perf::DiskIoError;
+    type CpuTimeError = WindowsError;
+
+    fn get_disk_io_counters(&self) -> Result<IoCounters, Self::DiskIoError> {
+        perf::get_disk_io_counters()
+    }
+
+    fn get_cpu_idle_time(&self) -> Result<f64, Self::CpuTimeError> {
+        perf::get_cpu_idle_time()
+    }
+}
+
+#[derive(Debug, Display)]
+pub enum WaitForIdleError<P>
+where
+    P: PerfProvider,
+{
+    #[display(fmt = "timed out waiting for CPU and disk to become idle")]
+    TimeoutError,
+    DiskIoError(P::DiskIoError),
+    CpuTimeError(P::CpuTimeError),
+}
+
+impl<P> Error for WaitForIdleError<P>
+where
+    P: PerfProvider,
+{
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            WaitForIdleError::TimeoutError => None,
+            WaitForIdleError::DiskIoError(ref e) => Some(e),
+            WaitForIdleError::CpuTimeError(ref e) => Some(e),
+        }
+    }
+}
+
+/// Wait for the CPU and disk to become idle.
+pub async fn cpu_and_disk_idle<P>(p: &P) -> Result<(), WaitForIdleError<P>>
+where
+    P: PerfProvider,
+{
+    let mut counters = p
+        .get_disk_io_counters()
+        .map_err(WaitForIdleError::DiskIoError)?;
+
+    for _ in 0..P::ATTEMPT_COUNT {
+        delay_for(Duration::from_millis(500)).await;
+
+        let new_counters = p
+            .get_disk_io_counters()
+            .map_err(WaitForIdleError::DiskIoError)?;
+        let idle = p
+            .get_cpu_idle_time()
+            .map_err(WaitForIdleError::CpuTimeError)?;
+
+        let delta_reads = new_counters.reads - counters.reads;
+        let delta_writes = new_counters.writes - counters.writes;
+
+        if idle >= 0.95f64 && delta_reads == 0 && delta_writes == 0 {
```

Comment:
Nit: slight preference for `0.95f64` being declared above as a constant.

Reply:
Also, should these tolerances be configurable? I don't know if the intention is to be able to run this on any arbitrary system running arbitrary software (for instance, to measure AV-affected startups), but it may be the case that requiring no reads or writes for a half second is untenable in some configurations.

Reply:
> 
> 
> Looks good. A general note that I might add though: I assume that we want to do this in order to reduce noise, which ultimately will allow us to see a signal in the results with fewer runs. However, I suspect that many users try to open Firefox very shortly after Windows is interactive. In this case, the test could have a distorted view of the impact that disk IO actually has on startup. It could thus be worth making this behavior (waiting on disk and CPU idle) configurable. Thoughts?

This was mostly just a port of the implementation in the [original startup recording](https://bitbucket.org/fqueze/frame-recording/src/3e356441e7d3d3737817e1bb7dec960977cb815e/fakeInput/fakeinput.py#lines-109) work by @fqueze.

> 
> 
> Also, should these tolerances be configurable? I don't know if the intention is to be able to run this on any arbitrary system running arbitrary software (for instance, to measure AV-affected startups), but it may be the case that requiring no reads or writes for a half second is untenable in some configurations.

It shouldn't be too hard to make these configurable, but the initial plan was to make this run in a single setup, that is, cold startup on reference hardware with no other restrictions.

Since the protocol is being rewritten in #12, I'll add the flag in a follow up PR to disable waiting for disk/CPU idle, but I'm going to leave the CPU idle% and disk idle time hardcoded until we actually need that to be configurable.